### ISSUE TITLE
Expose an API integrated with Python's AsyncIO?

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [grpc/grpc#6046]
I've talked about [this](https://docs.python.org/3/whatsnew/3.5.html#whatsnew-pep-492) before with @hsaliak (okay, to be fair, _he's talked with me_) but it's also worth having a larger discussion out in the open.

If this is something of interest for you, what are your expectations for it? Why would it be better than the asynchronous API currently in place for gRPC Python? What would it look like, and how would it behave (bonus points for sample code)?

**Comment by f0t0n** [comment]
For now it's not possible to use grpc with Python 3.5.1 and native async/await syntax. Current futures implementation is just a workaround.
Say I have an application which uses https://github.com/KeepSafe/aiohttp for asynchronous HTTP requests and https://github.com/aio-libs/aiopg/ to work with the database.

Say inside my service procedure I want to asynchronously call some HTTP resources with `aiohttp`, save something in the PostgreSQL using `aiopg` and get some cached data from Redis with `aioredis` and do some other I/O stuff asynchronously:


```py

import asyncio
import aiohttp
import aiopg
# ...


def save_in_the_db(url, data):
    # Save data in PostgreSQL database using aiopg
    pass


async def fetch(client, url):
    async with client.get(url) as resp:
        assert resp.status == 200
        return await resp.text()


async def my_grpc_endpoint(*args, **kwargs):
    with aiohttp.ClientSession() as client:
        data = await fetch(client, 'https://example.com/data.json')
        row = await save_in_the_db(data)
        # other async/await stuff
        return some_result
```

So to use async/await inside the grpc endpoint the endpoint itself should be a coroutine defined with `async def` syntax (or generator-based coroutine decorated with `@asyncio.coroutine`).

**Comment by leifurhauks** [comment]
@f0t0n 's example here is what I would expect as well.

**Comment by jtattermusch** [comment]
@nathanielmanistaatgoogle  just btw, C# API is currently using async/await and I've been getting good feedback on the clarity and usability of such API (but one thing to know is that async/await is a highly idiomatic thing in C#).

**Comment by lyschoening** [comment]
Is this something that is being actively worked on? It is possible to run an event loop in a separate thread and delegate all RPC calls to the event loop via `asyncio.run_coroutine_threadsafe()` in Python 3.5.1+, but it worries me that asyncio is not officially supported when it is obviously the way forward.

**Comment by nathanielmanistaatgoogle** [comment]
@lyschoening: just to be clear: this issue is about using gRPC Python with [Python 3.5-and-later's `async` and `await` syntax](https://docs.python.org/3/whatsnew/3.5.html#pep-492-coroutines-with-async-and-await-syntax), not [Python 3.4-and-later's `asyncio` module](https://docs.python.org/3.4/whatsnew/3.4.html#asyncio). We feel that [coroutines with `async` and `await` syntax](https://docs.python.org/3/whatsnew/3.5.html#pep-492-coroutines-with-async-and-await-syntax) "is obviously the way forward" and hope you agree (please respond in detail if you think we're mistaken about this!). This is recognized as something that will have to be done at some point. It is not actively being worked on, but it is actively being thought about.

**Comment by lyschoening** [comment]
Ok, that explains the odd phrasing of the issue. I posted here because it is the only issue that comes up when searching for "asyncio". If you support Servicers with coroutines there is no additional work required to support developers having Python 3.5 to use `async` and `await` instead of `@asyncio.coroutine` and `yield` in their method implementations, so my question was about whether there is activity towards support coroutines.

**Comment by nathanielmanistaatgoogle** [comment]
The current understanding is that because the gRPC Core operates (both service-side and client-side) by having a Python thread call long-running blocking functions on it, that supporting Python coroutines is going to require major design and architecture work in both gRPC Core and gRPC Python. We don't think it will be a case of the right ten lines of code in two places and the rest comes for free.

What might we be missing?

**Comment by lyschoening** [comment]
I agree this is the issue here. The Python implementation looks quite opaque to me, but is it not possible to avoid calling this long-running blocking function for each new request? Since this is done on the Python side by submitting to the thread pool I was hopeful that it could be refactored that some dedicated thread(s) block on communications from the server and forward the messages/requests to an asynchronous queue on a different thread that also runs the event loop.

**Comment by nathanielmanistaatgoogle** [comment]
@lyschoening: on the invocation-side, it is possible to avoid calling the Core's long-running blocking function for each RPC, and using a single shared thread to make that expensive function call during execution of arbitrarily many RPCs is what we do with the `grpc.Future`-returning and iterator-returning asynchronous client-side RPC methods available today. Unfortunately the iterators don't support registration of callbacks, but in the case of just response-unary RPC methods, could you use the `grpc.Future`s returned by your RPC invocations to implement the integration with an event loop that you describe? If not, what are the obstacles?

**Comment by virtuald** [comment]
Tornado and asyncio use a similar model for asynchronous operations, so the work I've done for integrating tornado with gRPC might be useful here. 

For integration with tornado, I've created a wrapping function that wraps gRPC's Rendevous object, and documented it in the [wiki](https://github.com/grpc/grpc/wiki/Integration-with-tornado-(python)). While a bit tedious, it's a good place to start and shows what's needed to make it work. The key is using the callback to thunk back to the main async processing thread.

I've also been able to get streaming working by modifying the Rendevous `_next` function, see #7910 .

Looking at both of these, it feels like having a pluggable way to specify how asynchronous operations are dispatched would make a lot of sense, and I think could be done without any major restructuring of gRPC core (until we start looking at performance, perhaps).

**Comment by virtuald** [comment]
A notional API for simple wrapping would involve:

* Adding a mechanism to allow overriding the `stub.fn.async` in unary calls to return a framework-specific Future, with the correct thunking to allow processing in the proper thread
* Adding a mechanism to allow overriding the iterable mechanisms in stream calls, and restructuring the iterable related functions (as in #7910) to be overridden

I haven't looked at what a gRPC server would require, but I suspect it's similar (use a platform-specific stub).

One could probably accomplish a lot of this via a factory that you pass to either the stub or the channel.

**Comment by nathanielmanistaatgoogle** [comment]
[We're going to try to use the magic of open source to address this enhancement request](https://groups.google.com/d/msg/grpc-io/RpkyqqQy8TU/UjL3ldKuAAAJ); please participate.

**Comment by mackeyja92** [comment]
+1 for asyncio

**Comment by amitripshtos** [comment]
Making the grpc python server using uvloop + asyncio will improve the performancec a lot of what it is at the moment. So, in my opinion we should make grpc use asyncio instead normal threading - and I'm willing to help in developing; let's make it happen!

**Comment by heri16** [comment]
+1 . Surprised that a google sponsored project does not care about performance.

**Comment by pokidovea** [comment]
Are there any changes?

**Comment by seglberg** [comment]
On the server side, this is what I've been using in the short term for executing our asyncio code: https://gist.github.com/seglberg/0b4487b57b4fd425c56ad72aba9971be

This monkey patch only works for Python 3.6+ since they did not add asynchronous generators until 3.6. 

Basically, you end up with two threads. One running the asyncio loop (used to execute the RPC methods) and another for managing the incoming requests. I then block the main thread with a `threading.Event()`

**Comment by costrouc** [comment]
@seglberg that worked as a drop in replacement for me on the server side. It was as simple as making each method async. 

How would this be done on the client side with async await? I see that each route has a future method call but it does not behave like asyncio futures... I see the add_done_callback but not sure if that is the right direction.

**Comment by seglberg** [comment]
I haven't spent much time on the client side, I've just been using the loop's `run_in_executor` when I need to make client calls from Python. Not completely ideal, but its a quick workaround if you don't mind spawning additional threads...

**Comment by mm-manu** [comment]
[Here](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/example/mnist_client.py) is an example of a gRPC client using the add_done_callback method.
I would like to avoid callbacks and prefer a method which waits until all concurrent requests are finished.
Please let me know if anyone knows how to do this.

**Comment by dstibrany** [comment]
@mm-manu Adapting virtuald's tornado example to asyncio seems to be working for me.
``` python
import asyncio

def _fwrap(f, gf):
    try:
        f.set_result(gf.result())
    except Exception as e:
        f.set_exception(e)

def fwrap(gf, loop=None):
    '''
        Wraps a GRPC result in a future that can be yielded by asyncio
        
        Usage::
        
            async def my_fn(param):
                result = await fwrap(stub.function_name.future(param, timeout))
        
    '''
    f = asyncio.Future()

    if loop is None:
        loop = asyncio.get_event_loop()

    gf.add_done_callback(lambda _: loop.call_soon_threadsafe(_fwrap, f, gf))
    return f
```

Then, as mentioned, you can do stuff like `result = await fwrap(stub.function_name.future(param, timeout))`

**Comment by VetoS88** [comment]
@seglberg, аs I understand this server for grpc requests, is the same asynchronous server, as a server based on aiohttp for http reqests?

**Comment by hubo1016** [comment]
I've developed a client library for grpc
https://github.com/hubo1016/aiogrpc

It uses the future() method to do async calls. But there are disadvantages: stream requests and respones are retrieved with separated threads. And also there is no server implementations.

I'm digging deeper these days. I found out that the Cython implementation is actually asynchronous, so there is really no need to use threads. But I need a stable Cython API (maybe documented) to work further.

Finally I would expect the asyncio interfaces to be just the corresponding replacement of the current interfaces:

1. future -> asyncio.Future
2. blocking interfaces -> coroutines
3. generators -> async generators

**Comment by mitar** [comment]
I opened #14876 to ask for the server-side asyncio Python implementation.

**Comment by decebal** [comment]
Hi @mitar , I wrote this article about my explorations not too long ago: [gRPC as a better alternative to GEARMAN](https://utopian.io/u/34589901), while [grpclib](https://github.com/vmagamedov/grpclib) might prove to be what you are looking for, if so please tell us about your experience with it.

**Comment by nickpro** [comment]
any news on this issue?

**Comment by ericgribkoff** [comment]
This is back on our radar but no concrete timeline yet. Will update here when we have any progress to report.

**Comment by Tehada** [comment]
bump

**Comment by Tehada** [comment]
3 years later. Still no changes! May be people can somehow help in implemetning it? I am ready to take part in it.

**Comment by FlorianLudwig** [comment]
@Tehada best way to help is to test and improve https://github.com/vmagamedov/grpclib

**Comment by gjcarneiro** [comment]
A pure Python grpclib is an interesting idea, but there are also advantages to just wrapping the standard grpc core C++ library for Python:
1. it is highly optimised;
2. it deals with things like multiple endpoints, health checks and round robin;
3. well tested.

It's just a pity grpclib core library is based on threads instead of asynchronous event loop (e.g. libuv).

Ideally, both efforts should continue in parallel.  There are pros and cons to both approaches, imho.

**Comment by heri16** [comment]
Any technical roadblocks as for a libuv-version of Grpclib?

**Comment by gjcarneiro** [comment]
But @hubo1016 said 
> I'm digging deeper these days. I found out that the Cython implementation is actually asynchronous, so there is really no need to use threads. But I need a stable Cython API (maybe documented) to work further.

So maybe I was wrong, maybe grpc core is already asynchronous internally, I don't know, I'm not a grpc developer :man_shrugging:

**Comment by szelenka** [comment]
I came across this implementation of a ThreadExecutor containing a single asyncio loop, which seems to function pretty well in converting the bi-directional streaming methods into Python async methods.:
https://gist.github.com/seglberg/0b4487b57b4fd425c56ad72aba9971be#file-grpc_asyncio-py

It allows you to use the standard grpc-tools way of generating the protoc stubs, and just modifying your implementation of it where async is needed (in contrast to grpclib).

However, it overwrites **grpc._server._unary_response_in_pool** and **grpc._server._stream_response_in_pool** to make this work in the async thread/loop. But I don't know much about the Python grpc package to know the implications of that.

I'm interested in others thoughts of the above approach towards adding asyncio to the Python gRPC.

**Comment by lidizheng** [comment]
AsyncIO support is under development. If you want to check the progress, please see https://github.com/grpc/grpc/projects/16. Most importantly, we are looking for contributors :)

**Comment by mitar** [comment]
I do not think it is stale, but it is in development, based on comment above?

**Comment by SaschaHeyer** [comment]
Is this still in progress?

**Comment by lidizheng** [comment]
This project is close to graduate from experimental stage https://github.com/grpc/grpc/pull/23240. We have been shipping parts by parts since `v1.25.0`. The API reference can be found at https://grpc.github.io/grpc/python/grpc_asyncio.html.

**Comment by MHDante** [comment]
Can this issue be closed thanks to #23240?

**Comment by lidizheng** [comment]
Sure. Thanks for reminding.
